MILESTONES.md

```diff
@@ -0,0 +1,152 @@
+    - Instead, all cursorless edits need to be applied as if they were real user
+      changes (they need to appear in the undo stack, etc)
+    - The JetBrains user should not need to save their file for Cursorless to
+      work.
+- We will change the sidecar extension to load this temporary file rather than
+  the real one.
+    - Every time the JetBrains state changes we will clobber the temporary file
+      and have the VS Code reload it.
+    - Long term, this is not ideal because Cursorless depends on receiving diffs
+      from the VS Code to manage its state.
+        - It can handle the file completely reloading, but this could cause the
+          hat map to change, which would prevent chaining multiple cursorless
+          commands within a single utterance.
+        - However, for this milestone chaining probably won't work in any
+          capacity, so will come back to this later.
+- JetBrains will record serial numbers/versions for every editor (or for the
+  state as a whole -- pokey's suggestion).
+    - We will attempt to tag the hats file with the serial number, so we can
+      tell quickly if we are viewing a stale hats file, and ignore it
+- We then need to identify when cursorless commands have run, and intelligently
+  apply that diff back to JetBrains.
+    - As above, it would be best if this was a precise diff, rather than "load
+      this alternate version of the file" so that we don't have to worry about
+      race conditions on JetBrains side (e.g., the user also making keyboard
+      changes at the same time).
+        - I imagine this would make the undo experience better as well.
+        - This is therefore significantly more important than sending smart
+          diffs to VS Code.
+    - VS Code provides existing watching APIs that we could used to listen to
+      changes, including those made by cursorless [1].
+        - These are nice because they return precise diffs, so we wouldn't have
+          to do that. We would need to filter out changes caused by us
+          synchronizing with JetBrains, however.
+        - One note: the Cursorless "execute command" promise guarantees that all
+          VS Code watchers that have been registered on the document are
+          notified _before it returns_, which is nice.
+        - This means, if we are the ones executing the Cursorless commands, we
+          could possibly listen for and collect all of the changes that were
+          made during the course of that command, returning them as a structured
+          object.
+    - Alternatively, we could run stuff after the `cursorless_command`
+      action finishes, and try to figure out what changed on the VS Code side.
+        - Since VS Code is working in a temporary file, we could have it save
+          its work and diff it against the original version.
+        - We could have it serialize its new cursor locations and just apply
+          them blindly to JetBrains. Again, risk of racing here.
```

Comment:
If sidecar extension includes most recent known JetBrains version in its dump, then in theory JetBrains could ignore the update / request a new one / wait for a new one?  Would need to think through details here.  I think having synchronization barrier of some sort is probably the only reliable answer here